### PR TITLE
support for custom game parameters for starting timers/enforcement

### PR DESCRIPTION
**Pull request by SeaGriff**

### CODE REVIEW

File: package-lock.json [hunk 1]
```diff
@@ -1,5 +1,5 @@
 {
-  "name": "cardservernode",
+  "name": "exceedgameserver",
```
Comment (daniel-k-taylor) [inline package-lock.json:2]: was there any particular reason you changed this? just curious

File: server.js [hunk 2]
```diff
@@ -49,6 +49,14 @@ var running_id = 1
 var running_match_id = 1
 var awaiting_match_room = null
 
+
+// Manages creating and joining a custom room.
```
Comment (daniel-k-taylor) [inline server.js:53]: Not a fan of this comment style.
If you want to break out the required parameters of join_room_json, that's fine, but it should be like:
// join_room_json required params:
//     game_version - blah blah
//     room_id - blah blah
// etc. for all params
Also, i probably want the comment inside the function for now.

database is a global here and is for logging stuff to the database.

File: server.js [hunk 3]
```diff
@@ -81,6 +89,7 @@ function join_custom_room(ws, join_room_json) {
     set_name(player, join_room_json)
   }
 
+  // Get the room id from the passed in json.
```
Comment (daniel-k-taylor) [inline server.js:92]: I don't particularly like comments like this that just say exactly what the code says.
var room_id = join_room_json.room_id.trim()
seems pretty clear

File: server.js [hunk 4]
```diff
@@ -98,20 +107,26 @@ function join_custom_room(ws, join_room_json) {
...
+    var starting_timer = join_room_json.starting_timer
+    var enforce_timer = join_room_json.enforce_timer
     var player = active_connections.get(ws)
     player.set_deck_id(deck_id)
     var success = false
+
+    // Check whether the player is joining an existing room.
     if (game_rooms.hasOwnProperty(room_id)) {
       const room = game_rooms[room_id]
+      // Check for Player/Room version mismatch.
```
Comment (daniel-k-taylor) [inline server.js:121]: Again, comments that just repeat what the code says exactly aren't super useful.
In this case, maybe plain "version" should be renamed to joining_player_version or something.

File: server.js [hunk 5]
```diff
@@ -98,20 +107,26 @@ function join_custom_room(ws, join_room_json) {
...
 
+    // Extract room settings from the passed in json.
     var deck_id = join_room_json.deck_id
+    var starting_timer = join_room_json.starting_timer
+    var enforce_timer = join_room_json.enforce_timer
     var player = active_connections.get(ws)
     player.set_deck_id(deck_id)
     var success = false
+
+    // Check whether the player is joining an existing room.
```
Comment (daniel-k-taylor) [inline server.js:118]: Can you put this comment inside the if statement and change it to:
// The room the player wants to join already exists.

File: server.js [hunk 6]
```diff
@@ -98,20 +107,26 @@ function join_custom_room(ws, join_room_json) {
...
+      // Check for Player/Room version mismatch.
       if (room.version != version) {
-        // Player/Room version mismatch.
         send_join_version_error(ws)
         return true
       }
       success = room.join(player)
     } else {
-      const new_room = new GameRoom(version, room_id, database)
+      // Start a new custom game room.
```
Comment (daniel-k-taylor) [inline server.js:128]: // The room doesn't exist, so start a new custom game room.

File: server.js [hunk 7]
```diff
@@ -98,20 +107,26 @@ function join_custom_room(ws, join_room_json) {
...
+      const new_room = new GameRoom(version, room_id, database, starting_timer, enforce_timer)
```
Comment (daniel-k-taylor) [inline server.js:129]: You need to update the other location that calls new GameRoom in this file in create_new_match_room().
Either pass in some default params or update GameRoom constructor to have default params.

File: gameroom.js [hunk 8]
```diff
@@ -182,6 +184,8 @@ class GameRoom {
         player2_name: this.players[1].name,
         player2_id: this.players[1].id,
         player2_deck_id: this.players[1].deck_id,
+        starting_timer: this.starting_timer,
+        enforce_timer: this.enforce_timer
```
Comment (daniel-k-taylor) [inline gameroom.js:188]: Can you add one more timer parameter?
I want a "overtime_per_move" as well.
The idea being that if you run out of time, you still get this amount of time per move.
If you run out of that, then you really lose!
Default should be like 15 or 20 seconds, but you can add it to your settings too.

File: gameroom.js [hunk 9]
```diff
@@ -182,6 +184,8 @@ class GameRoom {
         player2_name: this.players[1].name,
         player2_id: this.players[1].id,
         player2_deck_id: this.players[1].deck_id,
+        starting_timer: this.starting_timer,
```
Comment (daniel-k-taylor) [inline gameroom.js:187]: Please add these to the
const matchData = {
as well.
Everything in that dictionary gets saved in the game log when it completes.

**Review by daniel-k-taylor — CHANGES_REQUESTED** [review]

File: package-lock.json (hunk 1 above)
Comment (SeaGriff) [reply package-lock.json:2]: I never looked at this file, whatever this change is was automated

File: server.js (hunk 3 above)
Comment (SeaGriff) [reply server.js:92]: I find it easier when scanning to read brief summaries of the purpose of each block of code, then only read the code itself if I have a reason to - it turns a function into a story you can skim. but I do not mind removing this

File: gameroom.js (hunk 8 above)
Comment (SeaGriff) [reply gameroom.js:188]: that is a great idea

File: server.js (hunk 3 above)
Comment (daniel-k-taylor) [reply server.js:92]: This particular comment isn't a summary of a block of code though, it's literally rewriting the next line in english.
I think if you want to put in a summary of the block of code, that's fine.
I just don't want like a comment explaining every single line when the line pretty much says that.

File: server.js (hunk 4 above)
Comment (SeaGriff) [reply server.js:121]: renamed version to player_join_version

File: package-lock.json (hunk 1 above)
Comment (SeaGriff) [reply package-lock.json:2]: Not sure what to do about this, pls advise

File: gameroom.js (hunk 8 above)
Comment (SeaGriff) [reply gameroom.js:188]: what do you think of phrasing it this way? Minimum time given to take turn: if your turn starts and you're below x seconds, your timer gets set to x seconds

File: package-lock.json (hunk 1 above)
Comment (daniel-k-taylor) [reply package-lock.json:2]: I think this is probably fine. I suspect it is just based on the git name, which I changed.

File: gameroom.js (hunk 8 above)
Comment (daniel-k-taylor) [reply gameroom.js:188]: I'd like to make it clear in the UI and settings that this is overtime time that occurs after your full timer runs out.

Comment (SeaGriff) [reply gameroom.js:188]: Oh, I meant making it actually work in the way my phrasing suggests, apologies. I went ahead and coded minimum time per turn before I saw this - but I'm happy to add in overtime instead, or as an additional option

**Review by daniel-k-taylor — APPROVED** [review]
When this merges to main it will kick off a task to update the server, so let's wait and do that late when nobody is on.

File: server.js [hunk 10]
```diff
@@ -98,20 +104,26 @@ function join_custom_room(ws, join_room_json) {
     // Add a prefix to the room id to indicate custom match.
     room_id = "custom_" + room_id
 
+    // Extract room settings from the passed in json.
     var deck_id = join_room_json.deck_id
+    var starting_timer = join_room_json.starting_timer
```
Comment (daniel-k-taylor) [inline server.js:109]: Can you do validation that these are all ints and also check if they're missing. if they fail either, can you set them to some good default values?

**Review by daniel-k-taylor — CHANGES_REQUESTED** [review]

**Pull request merged**